Fix #22001, Added reverse iteration in SVG Sanitizer and wrote unit tests for validation

mishrayush15 opened the pull request:
## Overview

<!--
READ ME FIRST:
Please answer *all* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #22001.
2. This PR does the following: 
* Fixes a bug that caused an **'Unsupported tags/attributes'** error when uploading an SVG on the Creator Dashboard page by reversing the iteration of SVG-Sanitizer.
* Adds **unit tests** with **multiple cases** to ensure the error does not occur again
3. (For bug-fixing PRs only) The original bug occurred because: The SVG-Sanitizer **(File Path: core/templates/services/svg-sanitizer.service.ts | Line Number: 201)** previously used **forward iteration**, which caused certain attributes to be skipped during validation. In a forward iteration over an array, when an element at the **n**th position is removed, the **(n+1)**th element shifts to the **n**th position due to array shrinking. This shift caused the newly moved element to be **skipped** by the sanitizer check.

## Essential Checklist

Please follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).

- [x] I have linked the issue that this PR fixes in the "Development" section of the sidebar.
- [x] I have checked the "Files Changed" tab and confirmed that the changes are what I want to make.
- [x] I have written tests for my code.
- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [x] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I can't assign them directly).


- [ ] A testing doc has been written: ... (ADD LINK) ...
- [ ] _(To be confirmed by the server admin)_ All jobs in this PR have been verified on a live server, and the PR is safe to merge.

## Proof that changes are correct


https://github.com/user-attachments/assets/3a5962bb-2220-48f7-8f2e-2382e6b62f84


## Proof that the error was reproduced before fixing the bug


https://github.com/user-attachments/assets/1e84be97-44e5-4632-bcb5-2c97c0ad9f28




## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

mishrayush15 committed:
Fix: SVG validation and added unit test for validation

mishrayush15 committed:
Fix: Removed unnecessary comments

mishrayush15 committed:
Merge branch 'develop' into update-backend-error-logs

mishrayush15 committed:
Merge branch 'develop' into update-backend-error-logs

mishrayush15 commented:
@seanlip, PTAL!

mishrayush15 committed:
Merge branch 'develop' into update-backend-error-logs

seanlip commented on core/templates/services/svg-sanitizer.service.spec.ts:
```diff
@@ -579,4 +579,73 @@ describe('SvgSanitizerService', () => {
+          ' <circle cx="50" cy="50" r="40"' +
+          ' stroke="black" stroke-width="3" fill="red" />' +
+          ' </svg>',
+        expectedSvgString:
```
Slightly curious -- why does this change the order of xmlns and width/height? Maybe leaving a comment to explain would be helpful.

mishrayush15 replied:
The order of attributes matters a lot. In the previous code (during forward iteration), some invalid attributes were correctly deleted, which was good. However, when the order of attributes changed, some invalid attributes were skipped during checking, which later caused an error. So, it was necessary to test the SVG code with different attribute orders to make sure that no invalid attributes are missed in the future.

seanlip replied:
I have no concern about the order of attributes changing, I'm mainly suggesting leaving a comment to explain why the order changes between the original and final strings (not between the different test cases).

mishrayush15 replied:
Got it! I've updated the comments to explain why the attribute order changes between the original and final strings.

seanlip replied:
No, you haven't. What you've written (and what you mentioned in your earlier comment on this thread) is not what I am asking. It is clear that the test cases should cover different scenarios, so you don't need to explain that.

What I am asking is: in this svgString the attrs are, in order: xmlns:sodipodi, xmlns:xlink, width, height, xmlns. In the expected string they are: xmlns:xlink, xmlns, width, height. Notice that, even after removing xmlns:sodipodi, the order of the remaining attributes is changed. That is what I am asking for a clarifying comment about. Please fix.

mishrayush15 replied:
Hi @seanlip, I now understand exactly what you're asking me to clarify. After researching, I found a few possible reasons why the attribute order changes after removing unwanted attributes:

1. When we manipulate the SVG using JavaScript's DOM methods, the browser may automatically reorder attributes in a standardized way. In other words, when the browser parses an SVG into the DOM, it may adjust the attribute order.
2. When outerHTML converts the DOM back to a string, it applies a consistent attribute order determined by the browser.

Please let me know if this explanation makes sense so that I can proceed with pushing the updated comments, or if any further clarification is needed.

mishrayush15 replied:
In addition to this, could you please clarify whether I should write a comment explaining why the order of attributes changes after sanitizing the SVG code, followed by a one-line comment for each test case? Or do you suggest any other specific changes?

seanlip replied:
Thanks @mishrayush15. Can you cite sources/proof for the reasons you gave?

To your other question, for this comment, it's fine to have a general note at the top. For the other comment I left, that should be a one-line comment for each test case. (I'm a bit confused about how to respond here because I'm not sure whether your "one-line comment for each test case" refers to this thread or my other comment; if the latter, you are commenting in the wrong thread.)

Note: in the future, if you have general questions that relate to the PR as a whole, then, to avoid confusion, leave a top-level comment in GitHub instead (in the box that shows up after you click "Review Changes").

mishrayush15 replied:
#### Hi @seanlip, hope you are doing good. I apologize for the delayed response as I was traveling. To answer your question regarding proof for the reasons I gave, here is a small demo I ran in my browser that shows what I meant when I said that the DOM is responsible for the ordering of the attributes in an SVG:
---

#### Code :
```javascript
const parser = new DOMParser();
const serializer = new XMLSerializer();

const svgString = `<svg width="100" height="100"
    xmlns="http://www.w3.org/2000/svg"
    xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
    xmlns:sodipodi="http://sodipodi.sourceforge.net/DTD/sodipodi-0.dtd"
    xmlns:xlink="http://www.w3.org/1999/xlink"
    xmlns:dc="http://purl.org/dc/elements/1.1/">
    <circle cx="50" cy="50" r="40" stroke="black" stroke-width="3" fill="red"/>
</svg>`;

const svgDoc = parser.parseFromString(svgString, "image/svg+xml");
const newSvgString = serializer.serializeToString(svgDoc.documentElement);

console.log(newSvgString);
```

#### Proof/Output :
![image](https://github.com/user-attachments/assets/fc6d4ce2-848d-4362-a6c2-31551f01c6b8)

**As seen in the output, the attribute order in the SVG has changed after parsing and serializing. Below are certain points that I want to add on top of it:**

1. When we convert an SVG string to a DOM document and then back to a string, the browser standardizes the attribute order.
2. This reordering happens automatically due to how browsers internally handle the SVG DOM structure.

This supports my point that **when we manipulate SVGs using JavaScript's DOM methods, the browser may adjust the attribute order in a standardized way**. Hope this clarifies my reasoning! Let me know if you need any further explanation.

seanlip replied:
Yup, that's fine. Please explain similarly in a comment so that other devs aren't confused, thanks.

mishrayush15 replied:
For sure, also just to confirm, can I commit the changes with all the comments you suggested?

seanlip commented on core/templates/services/svg-sanitizer.service.spec.ts:
```diff
@@ -579,4 +579,73 @@ describe('SvgSanitizerService', () => {
+  it('should remove unwanted/unnecessary attributes and return a safe SVG', () => {
+    const testCases = [
+      {
+        svgString:
```
Having some simple one-line comments above each test, along the lines of 

`// In this image, xmlns:rdf, xmlns:sodipodi, xmlns:dc are removed.`

would help make it easier for the reader to see what's going on.

mishrayush15 replied:
Yep, I will make sure to add a meaningful comment to explain what this test does, so other developers can understand it easily.

seanlip replied:
Sorry, it looks like there was a misunderstanding. By "each test", I meant each test case -- i.e. each svgString/expectedSvgString pair. The comment should be similar to what I mentioned above. Please fix.

seanlip reviewed (commented):
@mishrayush15 Looks pretty good -- nice videos! I left a few comments, but @kevintab95 is actually your reviewer here so I'll defer the rest of the review to him. Please assign him once you've addressed the comments I left. Thanks!

mishrayush15 committed:
Fix: Added meaningful comments

mishrayush15 commented:
@kevintab95, PTAL!

kevintab95 commented on core/templates/services/svg-sanitizer.service.ts:
```diff
@@ -198,7 +198,7 @@ export class SvgSanitizerService {
       if (tagsToBeRemoved.indexOf(nodeTagName) !== -1) {
         node.remove();
       } else {
-        for (let i = 0; i < node.attributes.length; i++) {
```
Could you please add a comment here explaining why iterating in reverse is required?

mishrayush15 replied:
Got it! I will add a comment explaining why reverse iteration is required here.

kevintab95 reviewed (changes requested):
Thanks @mishrayush15. Left a minor comment. PTAL.

mishrayush15 committed:
Fix: Added comments explaing the requirement of reverse iteration

mishrayush15 commented:
Hi @kevintab95, I've added all the necessary comments, PTAL!

kevintab95 reviewed (approved):
LGTM, thanks @mishrayush15!

kevintab95 commented:
Thanks @mishrayush15! BTW, just a note for future PRs, when addressing review comments, feel free to respond with "done" or similar (like you have done above) but leave the thread unresolved -- this way reviewers can see relevant previous discussion that happened.

mishrayush15 commented:
> Thanks @mishrayush15! BTW, just a note for future PRs, when addressing review comments, feel free to respond with "done" or similar (like you have done above) but leave the thread unresolved -- this way reviewers can see relevant previous discussion that happened.

I apologize for the mistake! I'll make sure to leave the comments unresolved in future PRs. Thanks a lot for your valuable time!

mishrayush15 committed:
Merge branch 'develop' into update-backend-error-logs

mishrayush15 committed:
Fix: Updated comments for SVG Sanitizer Unit Test

mishrayush15 committed:
Fix: Improve comments in SVG Sanitizer unit test

mishrayush15 commented:
@seanlip, PTAL!

mishrayush15 committed:
Merge branch 'develop' into update-backend-error-logs

mishrayush15 committed:
Merge branch 'develop' into update-backend-error-logs

seanlip reviewed (approved):
Thanks! LGTM.

seanlip commented:
Just a note, please follow these instructions if CI checks fail on your PR -- thanks. https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR

mishrayush15 committed:
Merge branch 'develop' into update-backend-error-logs

mishrayush15 commented:
> Just a note, please follow these instructions if CI checks fail on your PR -- thanks. https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR

Hello, @seanlip. I have fixed all the CI failures, PTAL!
